os/fs/smartfs/smartfs_utils.c

```diff
@@ -1716,139 +1963,263 @@ int smartfs_countdirentries(struct smartfs_mountpt_s *fs, struct smartfs_entry_s
+	*(uint16_t *)chainheader->used = last_data;
+	*(uint16_t *)chainheader->nextsector = SMARTFS_ERASEDSTATE_16BIT;
+
+	/* New file length = requested length. Filepos is already at position 'length' */
+	sf->entry.datlen = length;
+	smartfs_sync_internal(fs, sf);
+#else
+	smartfs_setbuffer(&readwrite, sf->currsector, 0, fs->fs_llformat.availbytes, (FAR uint8_t *)fs->fs_rwbuffer);
+	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Failed to read file sector no. %d, error = %d\n", nextsector, ret);
+		return ret;
+	}
 
-	nextsector = entry->firstsector;
-	header = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+	chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+	memset(&fs->fs_rwbuffer[offset], CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes - offset);
+	*(uint16_t *)chainheader->used = last_data;
+	*(uint16_t *)chainheader->nextsector = SMARTFS_ERASEDSTATE_16BIT;
 
-	while (nextsector != SMARTFS_ERASEDSTATE_16BIT) {
-		/* Read the next sector's header into our buffer */
+	sf->entry.datlen = length;
+	ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Unable to write to sector %d while shrinking file, error = %d\n", readwrite.logsector, ret);
+		return ret;
+	}
+#endif
 
-		smartfs_setbuffer(&readwrite, nextsector, 0, sizeof(struct smartfs_chain_header_s), (uint8_t *)fs->fs_rwbuffer);
+	nextsector = SMARTFS_NEXTSECTOR(chainheader);
+	chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+
+	while (nextsector != SMARTFS_ERASEDSTATE_16BIT) {
+		smartfs_setbuffer(&readwrite, nextsector, 0, sizeof(struct smartfs_chain_header_s), (FAR uint8_t *)fs->fs_rwbuffer);
 		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
 		if (ret < 0) {
-			fdbg("Error reading sector %d header\n", nextsector);
+			fdbg("Unable to read sector %d, error = %d\n", readwrite.logsector, ret);
+			return ret;
+		}
+		sector = SMARTFS_NEXTSECTOR(chainheader);
+		ret = FS_IOCTL(fs, BIOC_FREESECT, (unsigned long)nextsector);
+		if (ret < 0) {
+			fdbg("Unable to free sector %d to shrink file, error = %d\n", nextsector, ret);
 			return ret;
 		}
+		nextsector = sector;
+	}
+	sf->filepos = savefilepos;
 
-		/* Get the next chained sector */
+	return OK;
+}
 
-		sector = SMARTFS_NEXTSECTOR(header);
+/****************************************************************************
+ * Name: smartfs_extendfile
+ *
+ * Description:
+ *   Zero-extend the length of a regular file to 'length'.
+ *
+ ****************************************************************************/
 
-		/* If this is the 1st sector of the file, then just overwrite
-		 * the sector data with the erased state value.  The underlying
-		 * SMART block driver will detect this and release the old
-		 * sector and create a new one with the new (blank) data.
-		 */
+int smartfs_extendfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
```

Comment:
Please see below, it is what I thought before.
```
static ssize_t smartfs_write(FAR struct file *filep, const char *buffer, size_t buflen)
{
	/* Write data in the middle of written data */
	while ((sf->filepos < sf->entry.datlen) && (buflen > 0)) {
		buflen -= readwrite.count;
		...
	}
	/* Now append remained data */
	bytewritten = smartfs_append_data(fs, buffer, buflen);
	ret = bytewritten;
	return ret;
}

int smartfs_extendfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
{
	int bytewritten = 0;
	int writesize;
	int ret;
	int data_len = fs->fs_llformat.availbytes - sizeof(struct smartfs_chain_header_s);
	uint8_t *buf = malloc(data_len);
	memset(buf, '\0', data_len);
	...
	smartfs_seek_internal(fs, sf, 0, SEEK_END);
	while (length > 0) {
		if (length > data_len) {
			writesize = data_len;
		} else {
			writesize = length;
		}
		ret = smartfs_append_data(fs, buf, remain);
		length -= ret;
		bytewritten += ret;
	}
	
	return bytewritten;
}
```

Reply:
Yes we can use same logic and common function for adding new sectors to file data chain.
Pending...

Reply:
Resolved